contracts/src/Summa.sol

```diff
@@ -193,28 +215,64 @@ contract Summa is Ownable {
 
     /**
      * @dev Submit commitment for a CEX
-     * @param proof ZK proof of the valid polynomial encoding
+     * @param snarkProof ZK proof of the valid polynomial encoding
+     * @param grandsumProof kzg proof of the grand sum
+     * @param totalBalances The array of total balances in the grand sum
      * @param timestamp The timestamp at which the CEX took the snapshot of its assets and liabilities
      */
     function submitCommitment(
-        bytes memory proof,
+        bytes memory snarkProof,
+        bytes memory grandsumProof,
+        uint256[] memory totalBalances,
         uint256 timestamp
     ) public onlyOwner {
-        require(proof.length == 5376, "Invalid proof length");
+        // Check input lengths
+        require(totalBalances.length > 0, "Invalid total balances length");
+        require(grandsumProof.length == (totalBalances.length * 0x40), "Invalid grandsum proof length");
+        require(snarkProof.length > grandsumProof.length, "Invalid snark proof length");
+        
         uint[] memory args = new uint[](1);
         args[0] = 1; // Workaround to satisfy the verifier (TODO remove after https://github.com/summa-dev/halo2-solidity-verifier/issues/1 is resolved)
         require(
-            polynomialEncodingVerifier.verifyProof(verifyingKey, proof, args),
-            "Invalid proof"
+            polynomialEncodingVerifier.verifyProof(verifyingKey, snarkProof, args),
+            "Invalid snark proof"
         );
-        // TODO slice the proof to get the ID and balance commitments
-        // require(
-        //     balanceCommitments.length == config.cryptocurrencies.length,
-        //     "Liability commitments and cryptocurrencies number mismatch"
-        // );
+        require(
+            totalBalances.length == config.cryptocurrencyNames.length,
+            "Liability commitments and cryptocurrencies number mismatch"
+        );
+
+        bytes memory slicedSnarkProof = new bytes(grandsumProof.length);
+        for (uint256 i = 0; i < grandsumProof.length; i++) {
+            slicedSnarkProof[i] = snarkProof[i + 64]; // Skip first 64 bytes, it's not for total balance.
```

Comment:
Solidity now supports array slicing, maybe we can try to take advantage of that? https://docs.soliditylang.org/en/v0.6.0/types.html#array-slices

Reply:
Saved 146,361 gas by applying the array slice to the `submitCommitment` method. Thanks!
It's updated at [eb067b9](https://github.com/summa-dev/summa-solvency/pull/267/commits/eb067b9ec0e15cf9672238cf2937fc429da3f057)

Reply:
Wow, I didn't even expect such an improvement, sounds great!